cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java

```diff
@@ -0,0 +1,226 @@
+    try {
+      TransactionRunners.run(transactionRunner, context -> {
+        StructuredTable table = context.getTable(TABLE_ID);
+        table.upsert(getRow(oauthProvider));
+
+      }, TableNotFoundException.class, InvalidFieldException.class);
+    } catch (TableNotFoundException e) {
+      throw new OAuthStoreException("OAuth provider table not found", e);
+    } catch (InvalidFieldException e) {
+      throw new OAuthStoreException("OAuth provider object fields do not match table", e);
+    }
+
+    String namespace = NamespaceId.DEFAULT.getNamespace();
+    try {
+      secureStoreManager.put(
+          namespace,
+          getClientCredsKey(oauthProvider.getName()),
+          GSON.toJson(oauthProvider.getClientCredentials()),
+          "OAuth client creds",
+          EMPTY_MAP);
+    } catch (IOException e) {
+      throw new OAuthStoreException("Failed to write to OAuth provider secure storage", e);
+    } catch (Exception e) {
+      throw new OAuthStoreException("Namespace \"" + namespace + "\" does not exist", e);
+    }
+  }
+
+  /**
+   * Get an OAuth provider.
+   *
+   * @param name name of {@link OAuthProvider} to read
+   * @throws OAuthStoreException if the read fails
+   */
+  public Optional<OAuthProvider> getProvider(String name) throws OAuthStoreException {
+    OAuthClientCredentials clientCreds;
+    try {
+      String clientCredsJson = new String(
+          secureStore.get(NamespaceId.DEFAULT.getNamespace(), getClientCredsKey(name)).get(),
+          StandardCharsets.UTF_8);
+      clientCreds = GSON.fromJson(clientCredsJson, OAuthClientCredentials.class);
+    } catch (IOException e) {
+      throw new OAuthStoreException("Failed to read from OAuth provider secure storage", e);
+    } catch (Exception e) {
+      return Optional.empty();
+    }
+
+    try {
+      return TransactionRunners.run(transactionRunner, context -> {
+        StructuredTable table = context.getTable(TABLE_ID);
+        Optional<StructuredRow> row = table.read(getKey(name));
+        return row.map(structuredRow -> fromRow(structuredRow, clientCreds));
+      }, TableNotFoundException.class, InvalidFieldException.class);
+    } catch (TableNotFoundException e) {
+      throw new OAuthStoreException("OAuth provider table not found", e);
+    } catch (InvalidFieldException e) {
+      throw new OAuthStoreException("OAuth provider object fields do not match table", e);
+    }
+  }
+
+  /**
+   * Write an OAuth refresh token for the given provider and credential.
+   *
+   * @param oauthProvider name of OAuth provider the refresh token is sourced from
+   * @param credentialId ID used to identify this credential
+   * @param token the {@link OAuthRefreshToken} to write
+   * @throws OAuthStoreException if the write fails
+   */
+  public void writeRefreshToken(String oauthProvider, String credentialId, OAuthRefreshToken token)
+      throws OAuthStoreException {
+    String namespace = NamespaceId.DEFAULT.getNamespace();
```

Comment:
Why it is using the Default namespace? Since it is an system app, shouldn't it be using the SYSTEM namespace instead?

Reply:
I tried to use SYSTEM, but I seem to get a "namespace not found" error. I'm not sure if there's some additional step I need to perform to make it available for the secure storage, or something else is at work.